[WIP][SPARK-22246][SQL] Improve performance of UnsafeRow, UnsafeArrayData, UnsafeMapData, and SpecificUnsafeRowJoiner by using MemoryBlock

kiszk opened the pull request:
## What changes were proposed in this pull request?

Waiting for merging #19222

This PR tries to use `MemoryBlock` in `UnsafeRow`, `UnsafeArray`, `UnsafeMap`, `SpecificUnsafeRowJoiner`. There are two advantages to use `MemoryBlock`
1. Increase flexibility of choosing memory type (e.g `byte[]`, `int[]`, `long[]`, or `DirectBuffer (offheap)`).
2. Improve runtime performance of memory access instead of using `Object`.

This PR can achieve **1.9x performance improvement** of `Read UnsafeArrayData` in [`UnsafeArrayDataBenchmark`](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/UnsafeArrayDataBenchmark.scala#L42).

Without this PR
```
OpenJDK 64-Bit Server VM 1.8.0_121-8u121-b13-0ubuntu1.16.04.2-b13 on Linux 4.4.0-22-generic
Intel(R) Xeon(R) CPU E5-2667 v3 @ 3.20GHz
Read UnsafeArrayData:                    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            845 /  863        993.3           1.0       1.0X
Double                                         935 / 1003        897.0           1.1       0.9X
```
With this PR
```
OpenJDK 64-Bit Server VM 1.8.0_121-8u121-b13-0ubuntu1.16.04.2-b13 on Linux 4.4.0-22-generic
Intel(R) Xeon(R) CPU E5-2667 v3 @ 3.20GHz
Read UnsafeArrayData:                    Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------
Int                                            477 /  486       1760.3           0.6       1.0X
Double                                         824 /  835       1018.0           1.0       0.6X
```


## How was this patch tested?

Use existing testsuites

kiszk committed:
introduce ByteArrayMemoryBlock, IntArrayMemoryBlock, LongArrayMemoryBlock, and OffheaMemoryBlock

kiszk committed:
OffHeapColumnVector uses UnsafeMemoryAllocator

kiszk committed:
UTF8String uses UnsafeMemoryAllocator

kiszk committed:
Platform.copymemory() in UsafeInMemorySorter uses new MemoryBlock

kiszk committed:
address review comments

kiszk committed:
fix test failures (e.g. String in UnsafeArrayData)

kiszk committed:
fix failures

kiszk committed:
minor update of UTF8String constructor

kiszk committed:
rename method name

kiszk committed:
remove unused code

kiszk committed:
update arrayEquals

kiszk committed:
rebase master

kiszk committed:
make more methods final

kiszk committed:
make fill method final in MemoryBlock

kiszk committed:
fix test failures

kiszk committed:
add testsuite

kiszk committed:
pass concrete type to the first argument of Platform.get*/put* to get better performance

kiszk committed:
rename methods related to hash

kiszk committed:
added methods for MemoryBlock

kiszk committed:
rebase with master

kiszk committed:
fix scala style error

SparkQA commented:
**[Test build #82634 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82634/testReport)** for PR 19472 at commit [`e51fb6a`](https://github.com/apache/spark/commit/e51fb6ae246a660ec2b0bdd1f6a53a0987d4a063).

SparkQA commented:
**[Test build #82634 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82634/testReport)** for PR 19472 at commit [`e51fb6a`](https://github.com/apache/spark/commit/e51fb6ae246a660ec2b0bdd1f6a53a0987d4a063).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public abstract class MemoryBlock implements Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82634/
Test FAILed.

SparkQA commented:
**[Test build #82636 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82636/testReport)** for PR 19472 at commit [`0738359`](https://github.com/apache/spark/commit/0738359193f4c3a7e7cbeb85d1fc10967f06ff4f).

SparkQA commented:
**[Test build #82636 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82636/testReport)** for PR 19472 at commit [`0738359`](https://github.com/apache/spark/commit/0738359193f4c3a7e7cbeb85d1fc10967f06ff4f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82636/
Test FAILed.

SparkQA commented:
**[Test build #82642 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82642/testReport)** for PR 19472 at commit [`a814eb3`](https://github.com/apache/spark/commit/a814eb3f08085b09a16f336b36fba8da24e4f34a).

SparkQA commented:
**[Test build #82642 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82642/testReport)** for PR 19472 at commit [`a814eb3`](https://github.com/apache/spark/commit/a814eb3f08085b09a16f336b36fba8da24e4f34a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82642/
Test FAILed.

kiszk commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #82651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82651/testReport)** for PR 19472 at commit [`a814eb3`](https://github.com/apache/spark/commit/a814eb3f08085b09a16f336b36fba8da24e4f34a).

SparkQA commented:
**[Test build #82651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82651/testReport)** for PR 19472 at commit [`a814eb3`](https://github.com/apache/spark/commit/a814eb3f08085b09a16f336b36fba8da24e4f34a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82651/
Test FAILed.

SparkQA commented:
**[Test build #82670 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82670/testReport)** for PR 19472 at commit [`2dbfb8c`](https://github.com/apache/spark/commit/2dbfb8c2a7ffe19f6bd2b60a8fe40c9c97ce46e9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82670/
Test PASSed.

SparkQA commented:
**[Test build #82756 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82756/testReport)** for PR 19472 at commit [`150e0a3`](https://github.com/apache/spark/commit/150e0a30ac4ed11c783d62d47c4404c854b03dd9).

kiszk committed:
use MemoryBlock in Murmur3 for performance reason

kiszk committed:
fix typo in comment

SparkQA commented:
**[Test build #82756 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/82756/testReport)** for PR 19472 at commit [`150e0a3`](https://github.com/apache/spark/commit/150e0a30ac4ed11c783d62d47c4404c854b03dd9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/82756/
Test PASSed.

kiszk commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #83166 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83166/testReport)** for PR 19472 at commit [`150e0a3`](https://github.com/apache/spark/commit/150e0a30ac4ed11c783d62d47c4404c854b03dd9).

SparkQA commented:
**[Test build #83166 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/83166/testReport)** for PR 19472 at commit [`150e0a3`](https://github.com/apache/spark/commit/150e0a30ac4ed11c783d62d47c4404c854b03dd9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/83166/
Test PASSed.

kiszk committed:
address review comment

kiszk committed:
rebase with master

kiszk committed:
fix failures

kiszk committed:
fix failures in ArrowColumnVectorSuite and FeatureHasherSuite

kiszk committed:
address review comments

rename LongArrayMemoryBlock to OnHeapMemoryBlock

remove intArrayMemoryBlock

separate copyMemory to copyFrom and writeTo and make them non-static

add assertion for range check

kiszk committed:
address review comments

kiszk committed:
fix test failures

kiszk committed:
fix test failures

kiszk committed:
remove MemoryBlock offset at caller site for allocate() and copyMemory()

kiszk committed:
provide generic copyFrom() and writeTo() in MemoryBlock

remove MemoryBlock offset at caller site for copyFrom() and writeTo()

kiszk committed:
address review comments

kiszk committed:
address review comments

kiszk committed:
fix indent

kiszk committed:
remove assert for performance

kiszk committed:
address review comments

kiszk committed:
fix compilation failure

kiszk committed:
address review comment

kiszk committed:
fix test failures

fix incorrect asserion

kiszk committed:
fix test failures

kiszk committed:
reduce duplicated code in hash

kiszk committed:
address review comments

kiszk committed:
address review comments

kiszk committed:
add review comment

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1261/
Test PASSed.

SparkQA commented:
**[Test build #87941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87941/testReport)** for PR 19472 at commit [`158e37c`](https://github.com/apache/spark/commit/158e37c703a6b4c401178b12ba0adaa0cc6baeb1).

kiszk committed:
add review comments

SparkQA commented:
**[Test build #87941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87941/testReport)** for PR 19472 at commit [`158e37c`](https://github.com/apache/spark/commit/158e37c703a6b4c401178b12ba0adaa0cc6baeb1).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87941/
Test FAILed.

kiszk commented:
Retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1264/
Test PASSed.

SparkQA commented:
**[Test build #87944 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87944/testReport)** for PR 19472 at commit [`158e37c`](https://github.com/apache/spark/commit/158e37c703a6b4c401178b12ba0adaa0cc6baeb1).

SparkQA commented:
**[Test build #87944 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87944/testReport)** for PR 19472 at commit [`158e37c`](https://github.com/apache/spark/commit/158e37c703a6b4c401178b12ba0adaa0cc6baeb1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87944/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1266/
Test PASSed.

SparkQA commented:
**[Test build #87946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87946/testReport)** for PR 19472 at commit [`f9fe320`](https://github.com/apache/spark/commit/f9fe320530ca26c5c6da5ba9e69abf9b9ffa1295).

SparkQA commented:
**[Test build #87946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87946/testReport)** for PR 19472 at commit [`f9fe320`](https://github.com/apache/spark/commit/f9fe320530ca26c5c6da5ba9e69abf9b9ffa1295).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public abstract class MemoryBlock implements Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87946/
Test FAILed.

kiszk committed:
address review comments

kiszk committed:
address review comments

kiszk committed:
address review comments

kiszk committed:
address review comment

kiszk committed:
address review comments

kiszk committed:
address review comment

kiszk committed:
Use MemoryBlock in UnsafeRow, UnsafeArrayData, UnsafeMapData, and SpecificUnsafeRowJoiner

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #88479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88479/testReport)** for PR 19472 at commit [`d09c629`](https://github.com/apache/spark/commit/d09c6295aaca56e7940906b1ca60131f9a30ee9b).

SparkQA commented:
**[Test build #88479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88479/testReport)** for PR 19472 at commit [`d09c629`](https://github.com/apache/spark/commit/d09c6295aaca56e7940906b1ca60131f9a30ee9b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public abstract class MemoryBlock implements Serializable `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88479/
Test FAILed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15789/
Test PASSed.
